# Merge context

In [1]:
import pandas as pd
import json

In [2]:
# depth of this file in the project
file_depth = '../..'

In [3]:
with open(file_depth + '/config/data_25m_config.json') as config_file:
    config = json.load(config_file)

In [4]:
new_context_path = config['new_context_path']
movies_with_context = pd.read_csv(file_depth + new_context_path + 'movies_with_context.csv', encoding="UTF-8")
movies_with_context.reset_index(drop=True)
movies_with_context

movieId                               title  movieYear titleType  \
0            1                    Toy Story (1995)       1995     movie   
1            2                      Jumanji (1995)       1995     movie   
2            3             Grumpier Old Men (1995)       1995     movie   
3            4            Waiting to Exhale (1995)       1995     movie   
4            5  Father of the Bride Part II (1995)       1995     movie   
...        ...                                 ...        ...       ...   
62290   209157                           We (2018)       2018     movie   
62291   209159           Window of the Soul (2001)       2001     movie   
62292   209163                    Bad Poems (2018)       2018     movie   
62293   209169                 A Girl Thing (2001)       2001   tvMovie   
62294   209171      Women of Devil's Island (1962)       1962     movie   

       isAdult  runtimeMinutes  directors       actor  genreAction  \
0            0              81  nm0005124   nm0000741            0   
1            0             104  nm0002653   nm0001372            0   
2            0             101  nm0222043   nm0025908            0   
3            0             124  nm0001845   nm0001365            0   
4            0             106  nm0796124   nm0003028            0   
...        ...             ...        ...         ...          ...   
62290        0             100  nm1415482  nm10312234            0   
62291        0              73  nm0142504   nm1546270            0   
62292        0              97  nm2520391   nm0468514            0   
62293        0             237  nm0003022   nm0121700            0   
62294        0              86  nm0659992   nm0033980            1   

       genreAdult  ...  genreMystery  genreNews  genreReality-tv  \
0               0  ...             0          0                0   
1               0  ...             0          0                0   
2               0  ...             0          0                0   
3               0  ...             0          0                0   
4               0  ...             0          0                0   
...           ...  ...           ...        ...              ...   
62290           0  ...             0          0                0   
62291           0  ...             0          0                0   
62292           0  ...             0          0                0   
62293           0  ...             0          0                0   
62294           0  ...             0          0                0   

       genreRomance  genreSci-fi  genreShort  genreSport  genreThriller  \
0                 0            0           0           0              0   
1                 0            0           0           0              0   
2                 1            0           0           0              0   
3                 1            0           0           0              0   
4                 0            0           0           0              0   
...             ...          ...         ...         ...            ...   
62290             0            0           0           0              0   
62291             0            0           0           0              0   
62292             0            0           0           0              0   
62293             1            0           0           0              0   
62294             0            0           0           0              0   

       genreWar  genreWestern  
0             0             0  
1             0             0  
2             0             0  
3             0             0  
4             0             0  
...         ...           ...  
62290         0             0  
62291         0             0  
62292         0             0  
62293         0             0  
62294         0             0  

[62295 rows x 36 columns]

In [5]:
new_context_ratings = config['new_context_ratings']
ratings = pd.read_csv(file_depth + new_context_ratings, encoding="UTF-8")
ratings.reset_index(drop=True)
ratings

userId  movieId  rating   timestamp  ratingId
0              1      296     5.0  1147880044         0
1              1      306     3.5  1147868817         1
2              1      307     5.0  1147868828         2
3              1      665     5.0  1147878820         3
4              1      899     3.5  1147868510         4
...          ...      ...     ...         ...       ...
25000072  162541    50872     4.5  1240953372  25000072
25000073  162541    55768     2.5  1240951998  25000073
25000074  162541    56176     2.0  1240950697  25000074
25000075  162541    58559     4.0  1240953434  25000075
25000076  162541    63876     5.0  1240952515  25000076

[25000077 rows x 5 columns]

In [6]:
rating_context_data = pd.read_csv(file_depth + new_context_path + 'rating_context_data.csv', encoding="UTF-8")
rating_context_data

ratingId  day  isWeekday  season  partOfDay     holiday
0                0    3          1       1          3  no_holiday
1                1    3          1       1          2  no_holiday
2                2    3          1       1          2  no_holiday
3                3    3          1       1          3  no_holiday
4                4    3          1       1          2  no_holiday
...            ...  ...        ...     ...        ...         ...
25000072  25000072    2          1       1          4  no_holiday
25000073  25000073    2          1       1          4  no_holiday
25000074  25000074    2          1       1          4  no_holiday
25000075  25000075    2          1       1          4  no_holiday
25000076  25000076    2          1       1          4  no_holiday

[25000077 rows x 6 columns]

#### Create context data

In [7]:
context_data = ratings.drop(columns=['timestamp'])
context_data = context_data.merge(rating_context_data, on='ratingId', how='left')
context_data = context_data.merge(movies_with_context, on='movieId', how='left')

context_data

userId  movieId  rating  ratingId  day  isWeekday  season  \
0              1      296     5.0         0    3          1       1   
1              1      306     3.5         1    3          1       1   
2              1      307     5.0         2    3          1       1   
3              1      665     5.0         3    3          1       1   
4              1      899     3.5         4    3          1       1   
...          ...      ...     ...       ...  ...        ...     ...   
25000072  162541    50872     4.5  25000072    2          1       1   
25000073  162541    55768     2.5  25000073    2          1       1   
25000074  162541    56176     2.0  25000074    2          1       1   
25000075  162541    58559     4.0  25000075    2          1       1   
25000076  162541    63876     5.0  25000076    2          1       1   

          partOfDay     holiday  \
0                 3  no_holiday   
1                 2  no_holiday   
2                 2  no_holiday   
3                 3  no_holiday   
4                 2  no_holiday   
...             ...         ...   
25000072          4  no_holiday   
25000073          4  no_holiday   
25000074          4  no_holiday   
25000075          4  no_holiday   
25000076          4  no_holiday   

                                                     title  ...  genreMystery  \
0                                      Pulp Fiction (1994)  ...           0.0   
1         Three Colors: Red (Trois couleurs: Rouge) (1994)  ...           0.0   
2         Three Colors: Blue (Trois couleurs: Bleu) (1993)  ...           0.0   
3                                       Underground (1995)  ...           0.0   
4                               Singin' in the Rain (1952)  ...           0.0   
...                                                    ...  ...           ...   
25000072                                Ratatouille (2007)  ...           0.0   
25000073                                  Bee Movie (2007)  ...           0.0   
25000074                    Alvin and the Chipmunks (2007)  ...           0.0   
25000075                           Dark Knight, The (2008)  ...           0.0   
25000076                                       Milk (2008)  ...           0.0   

         genreNews  genreReality-tv  genreRomance genreSci-fi genreShort  \
0              0.0              0.0           0.0         0.0        0.0   
1              0.0              0.0           0.0         0.0        0.0   
2              0.0              0.0           0.0         0.0        0.0   
3              0.0              0.0           0.0         0.0        0.0   
4              0.0              0.0           1.0         0.0        0.0   
...            ...              ...           ...         ...        ...   
25000072       0.0              0.0           0.0         0.0        0.0   
25000073       0.0              0.0           0.0         0.0        0.0   
25000074       0.0              0.0           0.0         0.0        0.0   
25000075       0.0              0.0           0.0         0.0        0.0   
25000076       0.0              0.0           0.0         0.0        0.0   

          genreSport  genreThriller  genreWar  genreWestern  
0                0.0            1.0       0.0           0.0  
1                0.0            0.0       0.0           0.0  
2                0.0            0.0       0.0           0.0  
3                0.0            0.0       1.0           0.0  
4                0.0            0.0       0.0           0.0  
...              ...            ...       ...           ...  
25000072         0.0            0.0       0.0           0.0  
25000073         0.0            0.0       0.0           0.0  
25000074         0.0            0.0       0.0           0.0  
25000075         0.0            0.0       0.0           0.0  
25000076         0.0            0.0       0.0           0.0  

[25000077 rows x 44 columns]

Check if there are any NaN values

In [8]:
nan_values = context_data[context_data.isna().any(axis=1)]
nan_values

userId  movieId  rating  ratingId  day  isWeekday  season  \
5921          43      720     3.5      5921    3          1       4   
7427          59      720     4.0      7427    2          1       1   
7855          61      720     5.0      7855    6          0       1   
13434        102      720     5.0     13434    5          1       1   
13604        103      720     5.0     13604    3          1       1   
...          ...      ...     ...       ...  ...        ...     ...   
24994379  162516     5264     3.5  24994379    3          1       1   
24994518  162516     5672     2.0  24994518    2          1       1   
24994630  162516     6003     3.0  24994630    1          1       2   
24994812  162516     6776     3.5  24994812    7          0       1   
24996480  162519      720     5.0  24996480    4          1       3   

          partOfDay     holiday title  ...  genreMystery genreNews  \
5921              4  no_holiday   NaN  ...           NaN       NaN   
7427              1  no_holiday   NaN  ...           NaN       NaN   
7855              4  no_holiday   NaN  ...           NaN       NaN   
13434             4  no_holiday   NaN  ...           NaN       NaN   
13604             2  no_holiday   NaN  ...           NaN       NaN   
...             ...         ...   ...  ...           ...       ...   
24994379          4  no_holiday   NaN  ...           NaN       NaN   
24994518          4  no_holiday   NaN  ...           NaN       NaN   
24994630          1  no_holiday   NaN  ...           NaN       NaN   
24994812          4  no_holiday   NaN  ...           NaN       NaN   
24996480          4  no_holiday   NaN  ...           NaN       NaN   

          genreReality-tv  genreRomance genreSci-fi genreShort  genreSport  \
5921                  NaN           NaN         NaN        NaN         NaN   
7427                  NaN           NaN         NaN        NaN         NaN   
7855                  NaN           NaN         NaN        NaN         NaN   
13434                 NaN           NaN         NaN        NaN         NaN   
13604                 NaN           NaN         NaN        NaN         NaN   
...                   ...           ...         ...        ...         ...   
24994379              NaN           NaN         NaN        NaN         NaN   
24994518              NaN           NaN         NaN        NaN         NaN   
24994630              NaN           NaN         NaN        NaN         NaN   
24994812              NaN           NaN         NaN        NaN         NaN   
24996480              NaN           NaN         NaN        NaN         NaN   

          genreThriller  genreWar  genreWestern  
5921                NaN       NaN           NaN  
7427                NaN       NaN           NaN  
7855                NaN       NaN           NaN  
13434               NaN       NaN           NaN  
13604               NaN       NaN           NaN  
...                 ...       ...           ...  
24994379            NaN       NaN           NaN  
24994518            NaN       NaN           NaN  
24994630            NaN       NaN           NaN  
24994812            NaN       NaN           NaN  
24996480            NaN       NaN           NaN  

[16608 rows x 44 columns]

There are, 16608 to be precise. That means there were some rows in movies_with_context with no values for matching "movieId" in dataframe ratings.

In [9]:
# for exaplme movieId 720 is not in the movies_with_context (but it is in the ratings)
movies_with_context[movies_with_context["movieId"] == 720]

Empty DataFrame
Columns: [movieId, title, movieYear, titleType, isAdult, runtimeMinutes, directors, actor, genreAction, genreAdult, genreAdventure, genreAnimation, genreBiography, genreChildren, genreComedy, genreCrime, genreDocumentary, genreDrama, genreFamily, genreFantasy, genreFilm-noir, genreHistory, genreHorror, genreImax, genreMusic, genreMusical, genreMystery, genreNews, genreReality-tv, genreRomance, genreSci-fi, genreShort, genreSport, genreThriller, genreWar, genreWestern]
Index: []

[0 rows x 36 columns]

In [10]:
print(f"Assumed number of rows after deletions: {context_data.shape[0]} - {len(nan_values)} = {context_data.shape[0] - len(nan_values)}")
context_data = context_data.dropna()
print('Length of ratings with context after dropping NaN values: ', len(context_data))

Assumed number of rows after deletions: 25000077 - 16608 = 24983469
Length of ratings with context after dropping NaN values:  24983469


But we also need to drop movies that were not rated at all.

In [ ]:
missing_movieIds = []

for movieId in movies_with_context['movieId']:
    if movieId not in context_data['movieId'].values:
        missing_movieIds.append(movieId)

In [ ]:
print(f"Number of movies with no rating:  {len(missing_movieIds)}")
print(f"Assumed number of movies after deletions: {movies_with_context.shape[0]} - {len(missing_movieIds)} = {movies_with_context.shape[0] - len(missing_movieIds)}")
movies_with_context = movies_with_context[~movies_with_context['movieId'].isin(missing_movieIds)]
print('Length of movies after dropping movies with no rating: ', len(movies_with_context))

Number of movies with no rating:  3372
Assumed number of movies after deletions: 62295 - 3372 = 58923
Length of movies after dropping movies with no rating:  58923


In [ ]:
movies_with_context.to_csv(file_depth + new_context_path + 'movies_with_context.csv', index=False)

In [11]:
context_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 24983469 entries, 0 to 25000076
Data columns (total 44 columns):
 #   Column            Dtype  
---  ------            -----  
 0   userId            int64  
 1   movieId           int64  
 2   rating            float64
 3   ratingId          int64  
 4   day               int64  
 5   isWeekday         int64  
 6   season            int64  
 7   partOfDay         int64  
 8   holiday           object 
 9   title             object 
 10  movieYear         float64
 11  titleType         object 
 12  isAdult           float64
 13  runtimeMinutes    float64
 14  directors         object 
 15  actor             object 
 16  genreAction       float64
 17  genreAdult        float64
 18  genreAdventure    float64
 19  genreAnimation    float64
 20  genreBiography    float64
 21  genreChildren     float64
 22  genreComedy       float64
 23  genreCrime        float64
 24  genreDocumentary  float64
 25  genreDrama        float64
 26  genreFamily      

In [12]:
context_data.to_csv(file_depth + new_context_path + 'context_data.csv', index=False)

Change float values to integers

In [13]:
del ratings
del rating_context_data
del movies_with_context
del context_data
del nan_values

In [14]:
dtypes = {
    'isAdult': int,
    'movieYear': int,
    'runtimeMinutes': int,
    'genreAction': int,
    'genreAdult': int,
    'genreAdventure': int,
    'genreAnimation': int,
    'genreBiography': int,
    'genreChildren': int,
    'genreComedy': int,
    'genreCrime': int,
    'genreDocumentary': int,
    'genreDrama': int,
    'genreFamily': int,
    'genreFantasy': int,
    'genreFilm-noir': int,
    'genreHistory': int,
    'genreHorror': int,
    'genreImax': int,
    'genreMusic': int,
    'genreMusical': int,
    'genreMystery': int,
    'genreNews': int,
    'genreReality-tv': int,
    'genreRomance': int,
    'genreSci-fi': int,
    'genreShort': int,
    'genreSport': int,
    'genreThriller': int,
    'genreWar': int,
    'genreWestern': int
}

In [15]:
new_context_path = config['new_context_path']
context_data = pd.read_csv(file_depth + new_context_path + 'context_data.csv', encoding="UTF-8", dtype=dtypes)
context_data.reset_index(drop=True)
context_data

userId  movieId  rating  ratingId  day  isWeekday  season  \
0              1      296     5.0         0    3          1       1   
1              1      306     3.5         1    3          1       1   
2              1      307     5.0         2    3          1       1   
3              1      665     5.0         3    3          1       1   
4              1      899     3.5         4    3          1       1   
...          ...      ...     ...       ...  ...        ...     ...   
24983464  162541    50872     4.5  25000072    2          1       1   
24983465  162541    55768     2.5  25000073    2          1       1   
24983466  162541    56176     2.0  25000074    2          1       1   
24983467  162541    58559     4.0  25000075    2          1       1   
24983468  162541    63876     5.0  25000076    2          1       1   

          partOfDay     holiday  \
0                 3  no_holiday   
1                 2  no_holiday   
2                 2  no_holiday   
3                 3  no_holiday   
4                 2  no_holiday   
...             ...         ...   
24983464          4  no_holiday   
24983465          4  no_holiday   
24983466          4  no_holiday   
24983467          4  no_holiday   
24983468          4  no_holiday   

                                                     title  ...  genreMystery  \
0                                      Pulp Fiction (1994)  ...             0   
1         Three Colors: Red (Trois couleurs: Rouge) (1994)  ...             0   
2         Three Colors: Blue (Trois couleurs: Bleu) (1993)  ...             0   
3                                       Underground (1995)  ...             0   
4                               Singin' in the Rain (1952)  ...             0   
...                                                    ...  ...           ...   
24983464                                Ratatouille (2007)  ...             0   
24983465                                  Bee Movie (2007)  ...             0   
24983466                    Alvin and the Chipmunks (2007)  ...             0   
24983467                           Dark Knight, The (2008)  ...             0   
24983468                                       Milk (2008)  ...             0   

         genreNews  genreReality-tv  genreRomance genreSci-fi genreShort  \
0                0                0             0           0          0   
1                0                0             0           0          0   
2                0                0             0           0          0   
3                0                0             0           0          0   
4                0                0             1           0          0   
...            ...              ...           ...         ...        ...   
24983464         0                0             0           0          0   
24983465         0                0             0           0          0   
24983466         0                0             0           0          0   
24983467         0                0             0           0          0   
24983468         0                0             0           0          0   

          genreSport  genreThriller  genreWar  genreWestern  
0                  0              1         0             0  
1                  0              0         0             0  
2                  0              0         0             0  
3                  0              0         1             0  
4                  0              0         0             0  
...              ...            ...       ...           ...  
24983464           0              0         0             0  
24983465           0              0         0             0  
24983466           0              0         0             0  
24983467           0              0         0             0  
24983468           0              0         0             0  

[24983469 rows x 44 columns]

In [16]:
context_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24983469 entries, 0 to 24983468
Data columns (total 44 columns):
 #   Column            Dtype  
---  ------            -----  
 0   userId            int64  
 1   movieId           int64  
 2   rating            float64
 3   ratingId          int64  
 4   day               int64  
 5   isWeekday         int64  
 6   season            int64  
 7   partOfDay         int64  
 8   holiday           object 
 9   title             object 
 10  movieYear         int32  
 11  titleType         object 
 12  isAdult           int32  
 13  runtimeMinutes    int32  
 14  directors         object 
 15  actor             object 
 16  genreAction       int32  
 17  genreAdult        int32  
 18  genreAdventure    int32  
 19  genreAnimation    int32  
 20  genreBiography    int32  
 21  genreChildren     int32  
 22  genreComedy       int32  
 23  genreCrime        int32  
 24  genreDocumentary  int32  
 25  genreDrama        int32  
 26  genreFamily 

#### Create prediction target (ratings) file

In [17]:
ratings_target = pd.DataFrame()
ratings_target['ratingId'] = context_data.index
ratings_target['rating'] = context_data['rating'].copy()
ratings_target

ratingId  rating
0                0     5.0
1                1     3.5
2                2     5.0
3                3     5.0
4                4     3.5
...            ...     ...
24983464  24983464     4.5
24983465  24983465     2.5
24983466  24983466     2.0
24983467  24983467     4.0
24983468  24983468     5.0

[24983469 rows x 2 columns]

Drop columns rating, ratingid and title, since they are not needed anymore

In [18]:
context_data = context_data.drop(columns=['rating', 'ratingId', 'title'])
context_data

userId  movieId  day  isWeekday  season  partOfDay     holiday  \
0              1      296    3          1       1          3  no_holiday   
1              1      306    3          1       1          2  no_holiday   
2              1      307    3          1       1          2  no_holiday   
3              1      665    3          1       1          3  no_holiday   
4              1      899    3          1       1          2  no_holiday   
...          ...      ...  ...        ...     ...        ...         ...   
24983464  162541    50872    2          1       1          4  no_holiday   
24983465  162541    55768    2          1       1          4  no_holiday   
24983466  162541    56176    2          1       1          4  no_holiday   
24983467  162541    58559    2          1       1          4  no_holiday   
24983468  162541    63876    2          1       1          4  no_holiday   

          movieYear titleType  isAdult  ...  genreMystery genreNews  \
0              1994     movie        0  ...             0         0   
1              1994     movie        0  ...             0         0   
2              1993     movie        0  ...             0         0   
3              1995     movie        0  ...             0         0   
4              1952     movie        0  ...             0         0   
...             ...       ...      ...  ...           ...       ...   
24983464       2007     movie        0  ...             0         0   
24983465       2007     movie        0  ...             0         0   
24983466       2007     movie        0  ...             0         0   
24983467       2008     movie        0  ...             0         0   
24983468       2008     movie        0  ...             0         0   

         genreReality-tv  genreRomance  genreSci-fi  genreShort  genreSport  \
0                      0             0            0           0           0   
1                      0             0            0           0           0   
2                      0             0            0           0           0   
3                      0             0            0           0           0   
4                      0             1            0           0           0   
...                  ...           ...          ...         ...         ...   
24983464               0             0            0           0           0   
24983465               0             0            0           0           0   
24983466               0             0            0           0           0   
24983467               0             0            0           0           0   
24983468               0             0            0           0           0   

          genreThriller  genreWar  genreWestern  
0                     1         0             0  
1                     0         0             0  
2                     0         0             0  
3                     0         1             0  
4                     0         0             0  
...                 ...       ...           ...  
24983464              0         0             0  
24983465              0         0             0  
24983466              0         0             0  
24983467              0         0             0  
24983468              0         0             0  

[24983469 rows x 41 columns]

In [19]:
context_data.to_csv(file_depth + new_context_path + 'context_data.csv', index=False)
ratings_target.to_csv(file_depth + new_context_path + 'ratings_target.csv', index=False)